processor/src/main/java/org/mapstruct/ap/model/source/selector/InheritanceTargetSelector.java

```diff
@@ -0,0 +1,107 @@
+/**
+ *  Copyright 2012-2015 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
+ *  copyright.txt file in the distribution for a full listing of all
+ *  contributors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.model.source.selector;
+
+import java.util.ArrayList;
+import java.util.List;
+import javax.lang.model.type.TypeMirror;
+
+import org.mapstruct.ap.model.common.Type;
+import org.mapstruct.ap.model.source.Method;
+
+/**
+ * Selects on inheritance distance, e.g. the amount of inheritance steps from the target type.
+ *
+ * @author Sjaak Derksen
+ */
+public class InheritanceTargetSelector implements MethodSelector {
```

Comment:
Btw. I checked with CDI, and it's the same for "producer methods" there, if there are several producers with an assignable return type, there is no form of "clever" resolution logic but one needs to resolve the ambiguity via qualifiers. Which makes me think we should do the same.